apis/Google.Cloud.Logging.V2/Google.Cloud.Logging.Log4Net/LruCache.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Logging.Log4Net
+{
+    class LruCache<K, V>
+    {
+        struct Item
+        {
+            public Item(LinkedListNode<K> node, V value)
+            {
+                Node = node;
+                Value = value;
+            }
+
+            public LinkedListNode<K> Node { get; }
+            public V Value { get; }
+        }
+
+        public LruCache(int capacity)
+        {
+            _capacity = capacity;
+            _dict = new Dictionary<K, Item>(capacity);
+            _order = new LinkedList<K>();
+        }
+
+        private readonly object _lock = new object();
+        private readonly int _capacity;
+        private readonly Dictionary<K, Item> _dict;
+        private readonly LinkedList<K> _order;
+
+        public V GetOrAdd(K key, Func<V> valueFn)
```

Comment:
I'd use a simple type.  string or object.   Using valueFn is adding uncertainty.   It could be taking too long or it could wait for other resource that it leads to deadlock in extreme case .